transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java

```diff
@@ -0,0 +1,317 @@
+            @Override
+            public void operationComplete(Future<C> future) throws Exception {
+                assert executor.inEventLoop();
+
+                if (future.isSuccess()) {
+                    originalPromise.setSuccess(future.getNow());
+                } else {
+                    // Something went wrong try to run pending acquire tasks.
+                    --acquiredChannelCount;
+
+                    assert acquiredChannelCount > 0;
+
+                    // Run the pending acquire tasks before notify the original promise so if the user would
+                    // try to acquire again from the ChannelFutureListener and the pendingAcquireCount is >=
+                    // maxPendingAcquires we may be able to run some pending tasks first and so allow to add
+                    // more.
+                    runTaskQueue();
+                    originalPromise.setFailure(future.cause());
+                }
+            }
+        });
+        if (acquiredChannelCount < maxConnections) {
+            ++acquiredChannelCount;
+
+            assert acquiredChannelCount > 0;
+
+            super.acquire(key, p);
+        } else {
+            if (++pendingAcquireCount > maxPendingAcquires) {
+                originalPromise.setFailure(FULL_EXCEPTION);
+            } else {
+                pendingAcquireQueue.offer(new AcquireTask<C, K>(key, p));
+                if (timeoutTask != null) {
+                    executor.schedule(timeoutTask, acquireTimeoutNanos, TimeUnit.NANOSECONDS);
+                }
+            }
+
+            assert pendingAcquireCount > 0;
+        }
+    }
+
+    @Override
+    public Future<Boolean> release(final C channel, final Promise<Boolean> promise) {
+        try {
+            final Promise<Boolean> p = executor.newPromise();
+            super.release(channel, p.addListener(new FutureListener<Boolean>() {
+
+                @Override
+                public void operationComplete(Future<Boolean> future) throws Exception {
+                    assert executor.inEventLoop();
+
+                    if (future.isSuccess()) {
+                        Boolean result = future.getNow();
+                        if (result == Boolean.TRUE) {
+                            --acquiredChannelCount;
+
+                            assert acquiredChannelCount >= 0;
+
+                            // Run the pending acquire tasks before notify the original promise so if the user would
+                            // try to acquire again from the ChannelFutureListener and the pendingAcquireCount is >=
+                            // maxPendingAcquires we may be able to run some pending tasks first and so allow to add
+                            // more.
+                            runTaskQueue();
+                        }
+                        promise.setSuccess(result);
+                    } else {
+                        promise.setFailure(future.cause());
```

Comment:
Lets be sure :) If need be we may have to add some additional state to know exactly what to do.  Also if we decrement `acquiredChannelCount` does that mean we should call `runTaskQueue`?